### PR TITLE
Requeue instance terminate message if host is not removed from the scheduler

### PR DESCRIPTION
**Pull request by lukeseawalker**
Requeue the instance terminate sqs message if one of the commands to remove
the host from the scheduler fail, in order to avoid the "dead node"
situation in SGE.

A "dead node" occurs when the instance is terminated but it's not
deleted from the scheduler configuration, because a job still references
it.

With this patch, if one of the commands to remove the host from the
scheduler configuration fail, then the sqs message containing the
instance terminate message is requeued in order to be reprocessed later.

When all the jobs referencing the terminated instance will be delete,
 then the host will be removed from the scheduler.

Test:
* when host cannot be removed, message is requeued
```
2019-02-07 18:22:41,631 INFO [sqswatcher:_poll_queue] event_type=autoscaling:EC2_INSTANCE_TERMINATE
2019-02-07 18:22:41,631 INFO [sqswatcher:_poll_queue] instance_id=i-00717b44e2edfc308
2019-02-07 18:22:41,637 INFO [sqswatcher:_remove_host] Removing hostname: ip-172-31-67-156
2019-02-07 18:22:41,637 INFO [sge:removeHost] Removing ip-172-31-67-156
2019-02-07 18:22:41,646 INFO [sge:removeHost] Host ip-172-31-67-156 is not administrative host
Attribute name ("*") and/or value ("ip-172-31-67-156") not found
"ip-172-31-67-156.ec2.internal" does not exist in "hostlist" of "hostgroup"
root@ip-172-31-66-16.ec2.internal modified "@allhosts" in host group list
Host object "ip-172-31-67-156" is still referenced in cluster queue "all.q".
2019-02-07 18:22:41,682 ERROR [sge:__run_sge_command] Failed to run ['/opt/sge/bin/lx-amd64/qconf', '-de', 'ip-172-31-67-156']

2019-02-07 18:22:41,682 INFO [sqswatcher:_poll_queue] Unable to remove host, requeuing autoscaling:EC2_INSTANCE_TERMINATE message 
```

* when host is removed
```
2019-02-07 18:23:41,987 INFO [sqswatcher:_poll_queue] event_type=autoscaling:EC2_INSTANCE_TERMINATE
2019-02-07 18:23:41,987 INFO [sqswatcher:_poll_queue] instance_id=i-00717b44e2edfc308
2019-02-07 18:23:41,994 INFO [sqswatcher:_remove_host] Removing hostname: ip-172-31-67-156
2019-02-07 18:23:41,994 INFO [sge:removeHost] Removing ip-172-31-67-156
2019-02-07 18:23:42,003 INFO [sge:removeHost] Host ip-172-31-67-156 is not administrative host
Attribute name ("*") and/or value ("ip-172-31-67-156") not found
"ip-172-31-67-156.ec2.internal" does not exist in "hostlist" of "hostgroup"
root@ip-172-31-66-16.ec2.internal modified "@allhosts" in host group list
root@ip-172-31-66-16.ec2.internal removed "ip-172-31-67-156.ec2.internal" from execution host list
root@ip-172-31-66-16.ec2.internal removed "ip-172-31-67-156.ec2.internal" from submit host list
2019-02-07 18:23:42,059 INFO [sqswatcher:_remove_host] Host ip-172-31-67-156 successfully removed from the cluster
2019-02-07 18:23:42,068 INFO [sqswatcher:_remove_host] Instance i-00717b44e2edfc308 successfully removed from the database
```

Signed-off-by: Luca Carrogu <carrogu@amazon.com>

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sqswatcher/plugins/sge.py [hunk 1]
```diff
@@ -19,28 +17,61 @@
...
+from sqswatcher.sqswatcher import GetSchedulerConfException
 
 log = logging.getLogger(__name__)
 
-def __runSgeCommand(command):
+
+def __is_host_in_command_output(command, hostname):
+    log.debug(repr(command))
+    _command = shlex.split(command)
+    log.debug(_command)
```
Comment (enrico-usai) [inline sqswatcher/plugins/sge.py]: We are printing the command two times. Is it intentional?

File: sqswatcher/plugins/sge.py [hunk 2]
```diff
@@ -19,28 +17,61 @@
...
+    output_has_hostname = False
+
+    try:
+        output = sub.check_output(_command,
```
Comment (enrico-usai) [inline sqswatcher/plugins/sge.py]: minor: the right indentation should be:
```
 output = sub.check_output(
     _command,
     env=dict(
         os.environ,
         SGE_ROOT='/opt/sge',
         PATH='/opt/sge/bin:/opt/sge/bin/lx-amd64:/bin:/usr/bin',
     ),
)
```

File: sqswatcher/plugins/sge.py [hunk 3]
```diff
@@ -19,28 +17,61 @@
 import socket
 import logging
 import shlex
+from sqswatcher.sqswatcher import RemoveHostFromSchedulerException
```
Comment (enrico-usai) [inline sqswatcher/plugins/sge.py]: We could think to move these exceptions in an utility module, imported by both `sqswatcher.py` and `sge.py` modules.

File: sqswatcher/plugins/sge.py [hunk 4]
```diff
@@ -95,33 +126,51 @@ def addHost(hostname, cluster_user, slots):
...
+        command = ("/opt/sge/bin/lx-amd64/qconf -dh %s" % hostname)
+        __run_sge_command(command, raise_exception=True)
+    else:
+        log.info('Host %s is not administrative host', hostname)
 
+    # Check if host is in all.q (qconf -sq all.q)
     # Purge hostname from all.q
     command = ("/opt/sge/bin/lx-amd64/qconf -purge queue '*' all.q@%s" % hostname)
-    __runSgeCommand(command)
+    __run_sge_command(command)
```
Comment (enrico-usai) [inline sqswatcher/plugins/sge.py]: Should we add the `__is_host_in_command_output` check for this command too?

Comment (lukeseawalker) [reply sqswatcher/plugins/sge.py]: I didn't add for two reasons:
* this command and the other below are non blocking, meaning the return code is always 0 also when you try to remove an host not present
* the format of the output returned by these two query command is not the same as the others and require a different parsing strategy

Comment (enrico-usai) [reply sqswatcher/plugins/sge.py]: Ok, thank you

File: sqswatcher/plugins/sge.py [hunk 5]
```diff
@@ -19,28 +17,64 @@
...
+        log.error("Failed to run %s\n" % command)
+        raise QueryConfigError
+
+    if output is not None:
+        # Expected output
+        # ip-172-31-66-16.ec2.internal
+        # ip-172-31-74-69.ec2.internal
+        match = filter(lambda x: hostname in x.split(".")[0], output.split("\n"))
+
+        if len(match) > 0:
```
Comment (enrico-usai) [inline sqswatcher/plugins/sge.py]: Note: `len(filter(...))` [is not supported](https://stackoverflow.com/questions/24291604/find-the-length-of-a-filter-object-in-python-3/24291630) by python3. 

http://python-future.org/compatible_idioms.html#filter-ifilter

File: sqswatcher/plugins/sge.py [hunk 6]
```diff
@@ -19,28 +17,64 @@
...
-    log.debug(repr(command))
+
+def _is_host_configured(command, hostname):
+    _command = shlex.split(command)
+    log.debug(_command)
+    host_configured = False
+
+    try:
+        output = sub.Popen(_command,
+                           stdout=sub.PIPE,
```
Comment (enrico-usai) [inline sqswatcher/plugins/sge.py]: Minor: the indentation is wrong

**Pull request merged**